executor/adapter.go

```diff
@@ -567,6 +567,12 @@ func (a *ExecStmt) handleNoDelayExecutor(ctx context.Context, e Executor) (sqlex
 		ctx = opentracing.ContextWithSpan(ctx, span1)
 	}
 
+	var err error
+	defer func() {
+		terror.Log(e.Close())
+		a.logAudit()
+	}()
+
```

Comment:
The `defer` statement is moved here because the following lines return 
```
errors.New("can not execute write statement when 'tidb_snapshot' is set")
```
and do not close the Executor.

Reply:
Can we move `e.Open()` to here for better management?

Reply:
We can't do that **due to the different logic between MustExec and MustQuery** (or you can call them `insert` and `select`).
For `insert`, `handleNoDelayExecutor` consume `Next()` and close the record set immediately. 

But for `select`, the record set is returned, that's why they are not in a function ....
We can not do those steps together in one function:

```
Open()
Next()
Close()
```